library/cartridge_set_instance_info.py

```diff
@@ -32,28 +30,58 @@ def get_instance_conf_section(app_name, instance_name, stateboard=False):
     return get_instance_id(app_name, instance_name, stateboard)
 
 
-def get_instance_work_dir(data_dir, app_name, instance_name, stateboard=False):
-    instance_id = get_instance_id(app_name, instance_name, stateboard)
-    return os.path.join(data_dir, instance_id)
-
-
 def get_instance_systemd_service(app_name, instance_name, stateboard=False):
     if stateboard:
         return '%s-stateboard' % app_name
     return '%s@%s' % (app_name, instance_name)
 
 
-def get_package_type(package_path):
-    if package_path is None:
-        return None
-    if package_path.endswith('.rpm'):
-        return 'rpm'
-    if package_path.endswith('.deb'):
-        return 'deb'
-    if package_path.endswith('.tar.gz'):
-        return 'tgz'
+def get_multiversion_app_code_dir(dist_dir, package_path):
+    package_basename = os.path.basename(package_path)
+
+    # get name and version
+    package_name_version, ext = os.path.splitext(package_basename)
+    if ext == '.gz':
+        package_name_version, _ = os.path.splitext(package_name_version)
+
+    app_code_dir = os.path.join(dist_dir, package_name_version)
+    return app_code_dir
+
+
+def get_multiversion_instance_code_dir(dist_dir, app_name, instance_name, stateboard=False):
+    instance_id = get_instance_id(app_name, instance_name, stateboard)
+    return os.path.join(dist_dir, instance_id)
 
-    raise Exception('Package of unsupported type is specified: %s' % package_path)
+
+def get_code_dirs_info(app_name, instance_name, instance_vars):
+    code_dirs_info = {}
+
+    if not instance_vars['cartridge_multiversion']:
+        app_code_dir = os.path.join(instance_vars['cartridge_dist_dir'], app_name)
+
+        code_dirs_info['app_code_dir'] = app_code_dir
+        code_dirs_info['instance_code_dir'] = app_code_dir
+        code_dirs_info['systemd_instance_code_dir'] = app_code_dir
+        code_dirs_info['systemd_stateboard_code_dir'] = app_code_dir
+    else:
+        code_dirs_info['app_code_dir'] = get_multiversion_app_code_dir(
+            instance_vars['cartridge_dist_dir'],
+            instance_vars['cartridge_package_path']
+        )
+
+        dist_dir = instance_vars['cartridge_dist_dir']
+        code_dirs_info['instance_code_dir'] = get_multiversion_instance_code_dir(
+            dist_dir, app_name, instance_name, instance_vars['stateboard'],
+        )
+
+        code_dirs_info['systemd_instance_code_dir'] = get_multiversion_instance_code_dir(
+            dist_dir, app_name, "%i",
+        )
+        code_dirs_info['systemd_stateboard_code_dir'] = get_multiversion_instance_code_dir(
+            dist_dir, app_name, None, stateboard=True,
+        )
```

Comment:
What about to make `instance_name=None` as default in `get_instance_id `, `get_multiversion_instance_code_dir` and others?
```suggestion
        code_dirs_info['systemd_stateboard_code_dir'] = get_multiversion_instance_code_dir(
            dist_dir, app_name, stateboard=True,
        )
```

Reply:
Yes, it's a good idea.